### ISSUE TITLE
~ on bools is confusing

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#5787]
``` chapel
  var t:bool(16) = true;
  var f:bool(16) = false;
  proc returnComplement(arg) {
    return ~arg;
  }
  writeln(returnComplement(t));
  writeln(returnComplement(f));
```

currently prints out
```
true
true
```

when one might expect it to print out
```
false
true
```

Related tests:
 *  types/scalar/bradc/bitNegateBool
 * types/scalar/bradc/bools2

Some discussion on this topic revealed these options:

1) Make bitwise operations on bool types (including `|` `&` `~`) result in compilation error.
 * programmers shouldn't be thinking of bools in terms of bit patterns / it's not what the bitwise operators are for
 * can lead to confusion if the element type ever changes to integer
 * could be confusing if coercion is involved

2) Make bitwise operations on bool types operate with 1-bit at a time; so `~` is equivalent to `!` and the and and or operators are equivalent to the logical versions, but without short circuiting.
 * arguably supports generic programming
 * these operations on bool values are unambiguous, so why not support them
 * traditionally, in C-like languages, `|` and `&` are available as non-short-circuiting equivalents to `||` and `&&`. If Chapel users care about this distinction, they might like to have both.

Related question: should it be possible to apply `||` `&&` or `!` to integral types?

**Comment by mppf** [comment]
Note, currently `!` is defined on integers, but `||` and `&&` are not.

**Comment by mppf** [comment]
Note, --llvm behaves differently than the C backend for some of these tests.

**Comment by vasslitvinov** [comment]
I support keeping integers and booleans separate: no bitwise operators on booleans and no logical operators on integers.

**Comment by gbtitus** [comment]
With respect to:

> ... traditionally, in C-like languages, | and & are available as non-short-circuiting equivalents to || and &&. If Chapel users care about this distinction, they might like to have both.

`|` and `&` are only "non-short-circuiting equivalents" for `||` and `&&` when programmers are very sloppy or values are very constrained.  `1 & 2` is 0, which C treats as logical false.  `1 && 2` is logical true.

**Comment by tmacd9** [comment]
Just to follow up on @gbtitus comments - if I write:

   var x : bool = (i < n) & (i > 0);

you are recommending this should be written as:

   var x : bool = (i < n) && (i > 0);

This could force someone to write:

   var x = f();
   var z = g() && x;

if they want both functions to be called instead of:

   var z = f() & g();

**Comment by gbtitus** [comment]
I think that falls under my qualifier "values are very constrained".  Because both operands of & in `(i < n) & (i > 0)` are boolean expressions, we're guaranteed they're both either 0 or 1, so there & could be used if the programmer insisted.  But that wouldn't be the case for, for example, f() & g() where those are functions returning int, unless there are additional guarantees that f() and g() always return either 0 or 1.  I do still think && would be more clear (and safe), though.

An additional difference, if my reading of the standard is correct: && has a sequence point between its left and right operands, while & does not.  So `f() && g()` guarantees that if `g()` is called then `f()` must have been called first.  On the other hand, `f() & g()` guarantees both are called, but doesn't guarantee the order in which they're called.

**Comment by tmacd9** [comment]
You are correct that there is a sequence point with `&&` and that the order for calling `f()` and `g()` is unspecified with the `&` operator.  That is also the case for `f() + g()` expressions (and other binary operators).  If both f and g return `bool ` values and there is no issue with calling them in either order, then this becomes a limitation.  Though, perhaps, not a serious limitation.

It becomes slightly more tedious if you want to code:

 ` while ( f() & g() ) { ... }
`
and make sure both f and g are called each time thru the loop.

**Comment by mppf** [comment]
Even with this discussion I still think that it'd be reasonable to allow both `&` and `&&` for expressions of `bool` type; where `&` would always evaluate LHS and RHS and `&&` would not.

At the same time, I don't think we should allow `&&` on non-bool types.

Also, I think that mixing `|` with a uint LHS and a bool RHS is interesting and useful. Sometimes you are making a bit mask or shifting a bit into place. Why can't you store the bit as a `bool`?

If you are setting the least-significant bit, it seems to me that `myUint = myUint | myBool` should do it. If you were setting another bit, say the bit in the 2s place, you'd write `myUint = myUint | ( myBool << 1)`. It seems clear enough to me to "promote" the bool to a uint in these cases. In fact, these cases work today (because of the coercion rules, I think).

In summary, I don't think it's reasonable to  *disallow* `|` " on bools".

**Comment by tmacd9** [comment]
On a similar note, there is no short circuit equivalent of `^` operations and that's reasonable with `bool` values.

There is a question  about `+ reduce A` where `A` is an  array of boolean values and the intent is to perform a popcount of true values,  Also, `| reduce A` can let you know if all values are false, and `& reduce A` can let you know if all values are trues.

**Comment by vasslitvinov** [comment]
If we don't allow `true + false`, we shouldn't allow `+ reduce` on an array of bools either. We should provide a `popcount` function instead.

**Comment by dmk42** [comment]
Gcc 7.1 has added a warning for using \~ on a bool, which we turn into an error with \-Wall.  We will not be able to pass a paratest with gcc 7.1 until we resolve this issue.

**Comment by mppf** [comment]
It seems fairly obvious to me that (2) is the only reasonable solution brought up. The other option (make it a compile error) would prevent certain idioms like | reduce. What I don't know is if the others involved in this discussion feel the same way.

**Comment by dmk42** [comment]
I would be fine with solution (2), converting bitwise operations into one-bit equivalents.

**Comment by vasslitvinov** [comment]
Both options make sense to me.

**Comment by mppf** [comment]
@vasslitvinov - if we chose (1), how would one write `| reduce A` on array of bools? And how would you write `^` on two bools, since there is no equivalent like `^^` (as `&` has `&&`) ?

**Comment by bradcray** [comment]
This seems to me like one of those topics that has exploded on us a bit.  The subject seems so deceptively simple ("Should we support ~ on bools?") but then it's gone off in a bunch of different directions.  Focusing on the question in the subject (which seems to be the one at hand due to gcc7 and llvm), it seems to me that we could choose not to support ~ on bools without making a similar call for & and |, and I'd be tempted to do so.  That said, I'm also OK with having ~ mean ! on bools, as long as it's well-documented.

**Comment by vasslitvinov** [comment]
@mppf - I think these can be worked out. Ex. `|| reduce A` and introduce `^^`. Although I am willing to admit that these are less-than-perfect.

**Comment by mppf** [comment]
@bradcray - even if we limited the discussion to `~` I think we don't agree. I think `~` on `bools` should just run `!` but it sounded like you wanted it to be an error (but were OK with the other answer). All I know is I seem to be more stubborn about this than others...

I can't tell if @vasslitvinov is playing devil's advocate or actually still prefers it being an error. 

@gbtitus also expressed support for not supporting `~` on `bool`. Greg, do you still feel that way?

@dmk42 indicated he was OK with `~` working as `!` on bools but hasn't said if he has a preference.

**Comment by dmk42** [comment]
I'll go along with whatever is decided so we can start using gcc 7.1.

**Comment by gbtitus** [comment]
I actually think the topic deserved explosion, because it seems not much (maybe no) harder to decide the larger language-level question than to just do a little fix to allow upgrading to gcc 7.1.

Perhaps I'm being over-purist but for me, predefining the bitwise operators on bools doesn’t particular make any more sense than predefining them on reals.  For whatever reason (perhaps too-long association with C), I view the int types as the only ones appropriate for the bitwise operators.  That said, I won’t throw a tantrum if we predefine `~` on bools.  It’ll just make me view Chapel’s `bool` type as less of a true boolean type and more just syntactic sugar for `int(1)`.

With a wink...
If we predefine `~&|^` on bools, for symmetry shouldn't we also predefine the other bitwise operators `<<` and `>>` on bools?  (And if we predefine `!` on ints, for symmetry shouldn’t we also predefine `&&` and `||` on ints?)

**Comment by dmk42** [comment]
Sorry, I didn't really mean to make gcc 7.1 the driver.  I just see both models as self-consistent and valid, and I think we should just pick a winner.

**Comment by bradcray** [comment]
I'm more open to supporting & and | on bools than ~ because we could define them for bools as meaning "non-short-circuiting and/or" rather than "bitwise" (though arguably, they're the same for a very crisp definition of bool in which the value is always 0x1 or 0x0).  Put another way, they add a capability that we wouldn't have otherwise.

In contrast, given that we have bool types like bool(8) and bool(64), it seems weird to support bitwise operators that would open the representation up to values other than 0x1 and 0x0.  So it seems weird to say that ~ represents bitwise complement on bools and it also seems weird to say that it's a 1-bit operation given that the bool's representation can be known to be a greater number of bits.  Meanwhile, there's no disadvantage to using ! on booleans (i.e., unlike & vs. &&, there's no real tradeoff between using ~ vs. !).

I suppose one could argue that supporting both would improve generic programming because a routine supporting any of these operators on either bools or integers, but I'd buy that argument better if we actually had a semi-compelling example in hand rather than just speculating one might exist.

**Comment by vasslitvinov** [comment]
I am with Brad in not needing ~ on booleans.

By contrast, we clearly need *logical* operators (AND and OR and perhaps XOR) on booleans, ideally both short-circuiting and not. Towards that end, I suggest that we do not think of | and & on booleans as "bit-wise". Just like we do not think of + on strings as doing some sort of arithmetic.

**Comment by bradcray** [comment]
In a meeting with Michael last week, I proposed that we remove the ~ overloads for bools for the time being on the argument that they didn't add capability and were potentially confusing, combined with the observation that it'd be easier to add the overload later if we found compelling use cases for doing so than to remove the overload later if we decided it was too wrong/confusing.

**Comment by dmk42** [comment]
Sounds good to me.  It also sounds like it would fit into what @gbtitus was advocating (he's out this week).

**Comment by noakesmichael** [comment]
+1



On May 23, 2017, at 2:25 PM, David Keaton <notifications@github.com<mailto:notifications@github.com>> wrote:


Sounds good to me. It also sounds like it would fit into what @gbtitus<https://github.com/gbtitus> was advocating (he's out this week).

**Comment by mppf** [comment]
Closed by PR #6300.

**Comment by ben-albrecht** [comment]
@mppf - Should the PR that closed this issue have made the `param bool` case an error as well?

https://github.com/chapel-lang/chapel/blob/master/modules/internal/ChapelBase.chpl#L516

A user (@npadmana) recently brought up the confusing result of a simple program:

```chpl
writeln(~true); // true
```

**Comment by mppf** [comment]
@ben-albrecht - PR #12023 - assuming testing goes ok
